File: android/src/main/kotlin/com/tapjoy/flutter/TapjoyOfferwallDiscover.kt
package com.tapjoy.flutter

import android.content.Context
import com.tapjoy.TJOfferwallDiscoverListener
import com.tapjoy.TJOfferwallDiscoverView
import com.tapjoy.TJError
import androidx.annotation.NonNull
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel.Result
import io.flutter.embedding.engine.plugins.FlutterPlugin
import io.flutter.plugin.common.MethodChannel.MethodCallHandler

class TapjoyOfferwallDiscover: MethodCallHandler {
    private val channelName = "tapjoy_offerwall_discover"
    companion object {
        var offerwallDiscover: TapjoyOfferwallDiscoverNativeView? = null
    }
    fun registerChannel(flutterPluginBinding: FlutterPlugin.FlutterPluginBinding) {
        val channel = ChannelManager.registerMethodChannel(flutterPluginBinding, channelName)
        channel.setMethodCallHandler { call, result ->
            onMethodCall(call, result)
        }
    }

    override fun onMethodCall(call: MethodCall, result: Result) {
        when (call.method) {
            "requestOWDContent" -> requestOWDContent(call, result)
            "clearOWDContent" -> clearOWDContent(call, result)
            else -> result.notImplemented()
        }
    }

    private fun requestOWDContent(@NonNull call: MethodCall, @NonNull result: Result) {
        TapjoyOfferwallPlugin.activity?.let {
            val owd = TapjoyOfferwallDiscoverNativeView(it)
            val placementName = call.argument("placementName") as String?
                ?: return result.error("ERROR", "placementName is null", null)
            owd.requestContent(placementName)
            offerwallDiscover = owd
        }
        return result.success(null)
    }

    private fun clearOWDContent(@NonNull call: MethodCall, @NonNull result: Result) {
        offerwallDiscover?.let {
            it.clearOWDContent()
            offerwallDiscover = null
        }
        return result.success(null)
    }
}

class TapjoyOfferwallDiscoverNativeView : TJOfferwallDiscoverView, TJOfferwallDiscoverListener {
    constructor(context: Context) : super(context)

    fun requestContent(placement: String) {
        setListener(this)
        requestContent(context, placement)
    }

    fun clearOWDContent() {
        super.clearContent()
    }

    override fun requestSuccess() {
        ChannelManager.sendEvent("requestDidSucceedForView", "Request succeeded for view.")
    }

    override fun requestFailure(error: TJError) {
        ChannelManager.sendError(error.code, "requestDidFailForView", error.message ?: "Request failed.")
    }

    override fun contentReady() {
        ChannelManager.sendEvent("contentIsReadyForView", "Content is ready for view.")
    }

    override fun contentError(error: TJError) {
        ChannelManager.sendError(error.code, "contentErrorForView", error.message ?: "Unknown content error")
    }
}

File: ios/Classes/TapjoyOfferwallDiscover.swift
import Flutter
import UIKit
import Tapjoy

class TapjoyOfferwallDiscover: NSObject{
    private let channelName = "tapjoy_offerwall_discover"
    var owdView: TapjoyOfferwallDiscoverNativeView?
    static let shared = TapjoyOfferwallDiscover()

    func registerChannel(with registrar: FlutterPluginRegistrar) {
         let channel = ChannelManager.shared.registerMethodChannel(forRegistrar: registrar,
                                                                        name: channelName)
        channel.setMethodCallHandler { [weak self] (call: FlutterMethodCall, result: @escaping FlutterResult) in
            self?.handleMethodCall(call: call, result: result)
        }
    }

    private func handleMethodCall(call: FlutterMethodCall, result: @escaping FlutterResult) {
            switch call.method {
            case "requestOWDContent":
                if let args = call.arguments as? [String: Any?] {
                    requestOWDContent(args, result: result)
                }
            case "clearOWDContent":
                    clearOWDContent(result: result)
            default:
                result(FlutterMethodNotImplemented)
            }
    }

    func requestOWDContent(_ args: [String: Any?], result: @escaping FlutterResult) {

        guard let placementName = args["placementName"] as? String else {
            result(FlutterError(code: "INVALID_ARGUMENT",
                                message: "Missing placementName",
                                details: nil))
            return
        }
        requestContent(placementName)
        result(nil)
    }

    func clearOWDContent(result: @escaping FlutterResult) {
        self.owdView?.clearContent()
        result(nil)
    }

    @objc func requestContent(_ placementName: String) {
        self.owdView = TapjoyOfferwallDiscoverNativeView()
        self.owdView?.delegate = self
        self.owdView?.requestContent(placementName)
    }

    private func sendEvent(_ eventName: String, _ message: String) {
        ChannelManager.shared.sendEvent(eventName, message)
    }

    private func sendError(_ eventName: String,  _ error: Error?) {
        let errorCode = (error as NSError?)?.code ?? -1
        ChannelManager.shared.sendError(errorCode, eventName, error?.localizedDescription ?? "Unknown Error")
    }
}

extension TapjoyOfferwallDiscover: TJOfferwallDiscoverDelegate {

    func requestDidSucceed(for view: TJOfferwallDiscoverView) {
        sendEvent("requestDidSucceedForView", "Request succeeded for view.")
    }

    func requestDidFail(for view: TJOfferwallDiscoverView, error: Error?) {
        sendError("requestDidFailForView", error)
    }

    func contentIsReady(for view: TJOfferwallDiscoverView) {
        sendEvent("contentIsReadyForView", "Content is ready for view.")
    }

    func contentError(for view: TJOfferwallDiscoverView, error: Error?) {
        sendError("contentErrorForView", error)
    }
}

